app/src/main/java/org/simple/clinic/encounter/EncounterRepository.kt

```diff
@@ -32,7 +34,32 @@ class EncounterRepository @Inject constructor(
   }
 
   override fun mergeWithLocalData(payloads: List<EncounterPayload>): Completable {
-    TODO("not implemented")
+    return Single.fromCallable {
+      payloads
+          .filter { payload ->
+            database.encountersDao()
+                .getOne(payload.uuid)
+                ?.syncStatus.canBeOverriddenByServerCopy()
+          }
+          .map(::payloadToEncounters)
+    }.flatMapCompletable(::saveMergedEncounters)
+  }
+
+  private fun payloadToEncounters(payload: EncounterPayload): EncounterAndObservations {
+    val bloodPressures = payload.observations.bloodPressureMeasurements.map { bps ->
+      bps.toDatabaseModel(syncStatus = DONE, encounterUuid = payload.uuid)
+    }
+    return EncounterAndObservations(encounter = payload.toDatabaseModel(DONE), bloodPressures = bloodPressures)
+  }
+
+  private fun saveMergedEncounters(records: List<EncounterAndObservations>): Completable {
+    return Completable.fromAction {
+      database
+          .bloodPressureDao()
+          .save(records.flatMap { it.bloodPressures })
+
+      database.encountersDao().save(records.map { it.encounter })
```

Comment:
Suggestion,
```kotlin
  private fun saveMergedEncounters(records: List<EncounterAndObservations>): Completable {
    return Completable.fromAction {
      val bloodPressures = records.flatMap { it.bloodPressures }
      val encounters = records.map { it.encounter }

      with(database) {
        bloodPressureDao().save(bloodPressures)
        encountersDao().save(encounters)
      }
    }
  }
```
Also, I feel uncomfortable saving the blood pressures and encounters outside a database transaction.

Reply:
I have changed to the following:
```
      database.openHelper.writableDatabase.inTransaction {
        with(database) {
          bloodPressureDao().save(records.flatMap { it.bloodPressures })
          encountersDao().save(records.map { it.encounter })
        }
      }

```
PTAL.

Reply:
```kotlin
with(database) {
  openHelper.writableDatabase.inTransaction {
    bloodPressureDao().save(bloodPressures)
    encountersDao().save(encounters)
  }
}
```